src/test/msgr/test_frames_v2.cc

```diff
@@ -0,0 +1,434 @@
+  bufferlist preamble_bl;
+  frame_bl->splice(0, frame_asm->get_preamble_onwire_len(), &preamble_bl);
+  frame_asm->disassemble_preamble(&preamble_bl, tag);
+
+  do {
+    size_t seg_idx = segment_bls->size();
+    segment_bls->emplace_back();
+
+    uint32_t onwire_len = frame_asm->get_segment_onwire_len(seg_idx);
+    if (onwire_len > 0) {
+      frame_bl->splice(0, onwire_len, &segment_bls->back());
+    }
+  } while (segment_bls->size() < frame_asm->get_num_segments());
+
+  bufferlist epilogue_bl;
+  uint32_t epilogue_onwire_len = frame_asm->get_epilogue_onwire_len();
+  if (epilogue_onwire_len > 0) {
+    frame_bl->splice(0, epilogue_onwire_len, &epilogue_bl);
+  }
+  frame_asm->disassemble_first_segment(&preamble_bl, segment_bls->data());
+  return frame_asm->disassemble_remaining_segments(segment_bls->data(),
+                                                   &epilogue_bl);
+}
+
+class RoundTripTestBase : public ::testing::TestWithParam<
+                              std::tuple<round_trip_instance_t, mode_t>> {
+protected:
+  RoundTripTestBase()
+      : m_tx_frame_asm(&m_tx_crypto, std::get<1>(GetParam()).is_r1),
+        m_rx_frame_asm(&m_rx_crypto, std::get<1>(GetParam()).is_r1),
+        m_header(make_bufferlist(std::get<0>(GetParam()).header_len, 'H')),
+        m_front(make_bufferlist(std::get<0>(GetParam()).front_len, 'F')),
+        m_middle(make_bufferlist(std::get<0>(GetParam()).middle_len, 'M')),
+        m_data(make_bufferlist(std::get<0>(GetParam()).data_len, 'D')) {
+    const auto& m = std::get<1>(GetParam());
+    if (m.is_secure) {
+      AuthConnectionMeta auth_meta;
+      auth_meta.con_mode = CEPH_CON_MODE_SECURE;
+      auth_meta.connection_secret.resize(64);
+      g_ceph_context->random()->get_bytes(auth_meta.connection_secret.data(),
+                                          auth_meta.connection_secret.size());
+      m_tx_crypto = ceph::crypto::onwire::rxtx_t::create_handler_pair(
+          g_ceph_context, auth_meta, /*new_nonce_format=*/m.is_r1,
+          /*crossed=*/false);
+      m_rx_crypto = ceph::crypto::onwire::rxtx_t::create_handler_pair(
+          g_ceph_context, auth_meta, /*new_nonce_format=*/m.is_r1,
+          /*crossed=*/true);
+    }
+  }
+
+  void check_frame_assembler(const FrameAssembler& frame_asm) {
+    const auto& [rti, m] = GetParam();
+    const auto& onwire_lens = rti.onwire_lens[m.is_r1 << 1 | m.is_secure];
+    EXPECT_EQ(rti.header_len + rti.front_len + rti.middle_len + rti.data_len,
+              frame_asm.get_frame_logical_len());
+    ASSERT_EQ(rti.num_segments, frame_asm.get_num_segments());
+    EXPECT_EQ(onwire_lens[0], frame_asm.get_preamble_onwire_len());
+    for (size_t i = 0; i < rti.num_segments; i++) {
+      EXPECT_EQ(onwire_lens[i + 1], frame_asm.get_segment_onwire_len(i));
+    }
+    EXPECT_EQ(onwire_lens[rti.num_segments + 1],
+              frame_asm.get_epilogue_onwire_len());
+    EXPECT_EQ(std::reduce(std::begin(onwire_lens), std::end(onwire_lens),
```

Comment:
The weird FTBS we were talking last week:

```
[ 89%] Building CXX object src/test/msgr/CMakeFiles/unittest_frames_v2.dir/test_frames_v2.cc.o
cd /work/ceph-1/build/src/test/msgr && /usr/bin/c++  -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D__linux__ -I/work/ceph-1/build/src/include -I/work/ceph-1/src -isystem /work/ceph-1/build/boost/include -isystem /work/ceph-1/build/include -isystem /work/ceph-1/src/xxHash -isystem /work/ceph-1/src/rapidjson/include -isystem /work/ceph-1/src/fmt/include -isystem /work/ceph-1/src/googletest/googlemock/include -isystem /work/ceph-1/src/googletest/googlemock -isystem /work/ceph-1/src/googletest/googletest/include -isystem /work/ceph-1/src/googletest/googletest -isystem /work/ceph-1/src/rocksdb/include -isystem /usr/include/fuse  -fno-omit-frame-pointer -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -rdynamic  -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wstrict-null-sentinel -Woverloaded-virtual -fno-new-ttp-matching -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -fstack-protector-strong -fdiagnostics-color=auto -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -O2 -g -DNDEBUG -fPIE   -DHAVE_CONFIG_H -D__CEPH__ -D_REENTRANT -D_THREAD_SAFE -D__STDC_FORMAT_MACROS -std=c++17 -o CMakeFiles/unittest_frames_v2.dir/test_frames_v2.cc.o -c /work/ceph-1/src/test/msgr/test_frames_v2.cc
In file included from /work/ceph-1/src/googletest/googletest/include/gtest/gtest.h:376,
                 from /work/ceph-1/src/test/msgr/test_frames_v2.cc:28:
/work/ceph-1/src/test/msgr/test_frames_v2.cc: In member function ‘void ceph::msgr::v2::RoundTripTestBase::check_frame_assembler(const ceph::msgr::v2::FrameAssembler&)’:
/work/ceph-1/src/test/msgr/test_frames_v2.cc:190:20: error: ‘reduce’ is not a member of ‘std’
     EXPECT_EQ(std::reduce(std::begin(onwire_lens), std::end(onwire_lens),
                    ^~~~~~
/work/ceph-1/src/test/msgr/test_frames_v2.cc:190:20: note: suggested alternative: ‘replace’
src/test/msgr/CMakeFiles/unittest_frames_v2.dir/build.make:62: recipe for target 'src/test/msgr/CMakeFiles/unittest_frames_v2.dir/test_frames_v2.cc.o' failed
make[2]: *** [src/test/msgr/CMakeFiles/unittest_frames_v2.dir/test_frames_v2.cc.o] Error 1
make[2]: Leaving directory '/work/ceph-1/build'
CMakeFiles/Makefile2:17780: recipe for target 'src/test/msgr/CMakeFiles/unittest_frames_v2.dir/all' failed
make[1]: *** [src/test/msgr/CMakeFiles/unittest_frames_v2.dir/all] Error 2
make[1]: Leaving directory '/work/ceph-1/build'
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

It's strange as `std::reduce` is available in C++17, `-std=c++17` is passed and, finally, the `numeric` header is in place. GCC version on this environment is 8.3.0.

Likely the reason is partial support for C++17. A snippet with `std::reduce` requires at least GCC 9.3.0, so maybe let's just workaround the issue and move to `std::accumulate` as proposed?

Reply:
Sure, I can do that.